Fix of 3.x series tutorial paragraph about error promotion of join call (#2458)
typelevel/cats-effect: docs/tutorial.md

}
```

However in most situations it is not advisable to handle fibers manually as
they are not trivial to work with. For example, if there is an error in a fiber

> the `join` call to that fiber will _not_ hint it, it will just 'swallow' the

Reviewer: Maybe I have just as much misconceptions about this stuff as the next person ... but doesn't it bubble up the error in the `Outcome`? It's not a loud as _raising_ the error for sure, but I wouldn't say it's not hinting it or swallowing it?

Reviewer 2: I'm in agreement, I think we should adjust the wording there.

Author: No problem on changing the paragraph :slightly_smiling_face: . But the expression 'swallow' was chosen purposefully. This code in CE3 won't show any error, it will show that last msg `All is fine`:
```scala
object TestStartJoin extends IOApp {

  def ok: IO[Unit] = IO(println("Waiting...")) >> IO.sleep(2 seconds) >> IO(println("Finished waiting"))
  def boom: IO[Unit] = IO.sleep(1 seconds) >> IO.raiseError(new Error("BOOM!"))

  override def run(args: List[String]): IO[ExitCode] =
    for {
      okFiber <- ok.start
      boomFiber <- boom.start
      _ <- okFiber.join
      _ <- boomFiber.join
      _ <- IO(println("All was fine"))
    } yield ExitCode.Success

}
```
Maybe 'swallows' is not the most fortunate expression, but I think the new paragraph should convey a clear warning that errors like the one shown above won't be raised.

Reviewer: @lrodero what do you think about my proposed wording in https://github.com/typelevel/cats-effect/pull/2458#discussion_r743157479?

Reviewer: Yes, I'm hesitant to say "swallows" because that would imply the error is gone...it's there in the outcome, which in this example is being discarded.

> the `join` call to that fiber will _not_ hint it, it will just 'swallow' the
> error. Also, the other fibers will keep running unaware of what happened.

Reviewer: ```suggestion
the `join` call to that fiber will _not_ raise it, it will return normally
and you must explicitly check the `Outcome` to see if it errored.
Also, the other fibers will keep running unaware of what happened.
```

Reviewer: @vasilmkd thoughts?

Author: LGTM! But I'm a bit concerned that the reader at this point does not know what `Outcome` is or where it comes from, so I'd change it to something like this:
```suggestion
the `join` call to that fiber will _not_ raise it, it will return normally and
you must explicitly check the `Outcome` instance returned by the `.join` call to
see if it errored. Also, the other fibers will keep running unaware of what
happened.
```

Reviewer: 👍 good point

> Cats-effect 3 provides additional `joinWith` or `joinWithNever` methods to make

Reviewer 3: ```suggestion
Cats Effect provides additional `joinWith` or `joinWithNever` methods to make
```

> sure at least that the error is promoted and give us a chance to deal with it.

Reviewer: ```suggestion
sure at least that the error is raised with the usual `MonadError` semantics (e.g., short-circuiting).
```

Unfortunately this is a complex task, as when we handle the error from a fiber
we must also consider if we should at least cancel the other running fibers. We

> can easily get ourselves trapped in a tangled mess of fibers to keep an eye of.

Reviewer: ```suggestion
can easily get ourselves trapped in a tangled mess of fibers to keep an eye on.
```

On top of that the error raised by a fiber is not promoted until the call to
`joinWith` or `.joinWithNever` is reached. So in our example above if
`consumerFiber` raises an error then that error will not 'reach' the main flow
until the producer fiber has finished. Note that in our example that should
never happen!  And even if the producer fiber did finish, it would have been
consuming resources for nothing.
 
In contrast `parMapN` does promote the error it finds to the caller _and_ takes
care of cancelling the remaining alive fibers. As a result `parMapN` is simpler

> to use, more concise, and ease to reason with. _Because of that, if possible,

Reviewer: ```suggestion
to use, more concise, and easier to reason about. _Because of that, if possible,
```

> to use, more concise, and easier to reason about. _Because of that, if possible,
> programmers should prefer to use higher level commands such as `parMapN` or
> `parSequence` to deal with fibers_.

Reviewer: ```suggestion
to use, more concise, and easier to reason about. _Because of that, unless you have some specific and
unusual requirements you should prefer to use higher level commands such as
`parMapN` or `parSequence` to work with fibers_.
```

}
```

However in most situations it is not advisable to handle fibers manually as
they are not trivial to work with. For example, if there is an error in a fiber
the `join` call to that fiber will _not_ raise it, it will return normally and
you must explicitly check the `Outcome` instance returned by the `.join` call to
see if it errored. Also, the other fibers will keep running unaware of what

> happened.
> Cats Effect provides additional `joinWith` or `joinWithNever` methods to make

Reviewer: ```suggestion
happened.

Cats Effect provides additional `joinWith` or `joinWithNever` methods to make
```

I prefer paragraphs to big blocks of text, but no biggie.

sure at least that the error is raised with the usual `MonadError` semantics (e.g., short-circuiting).

> Unfortunately this is a complex task, as when we handle the error from a fiber
> we must also consider if we should at least cancel the other running fibers. We

Reviewer: ```suggestion
Now that we are raising the error, an additional complexity is that we need to cancel
the other running fibers so as to propagate the short-circuiting semantics. We
```

Author: Hummm... sorry IMHO the new proposed text is less clear than the original :grimacing: . The wording in _An additional complexity is (...)_ looks a bit weird to me (though I must admit I'm not a native English speaker, by a long shot).

Reviewer: No worries! Is it just that part? I'm happy to drop it completely.

Reviewer: ```suggestion
Now that we are raising the error, we also need to cancel
the other running fibers so as to propagate the short-circuiting semantics. We
```

Reviewer: Or even:

```suggestion
Now that we are raising the error, we also need to cancel the other running fibers. We
```

Author: Nice one, it simplifies the text :+1:

can easily get ourselves trapped in a tangled mess of fibers to keep an eye on.
On top of that the error raised by a fiber is not promoted until the call to

> `joinWith` or `.joinWithNever` is reached. So in our example above if
> `consumerFiber` raises an error then that error will not 'reach' the main flow
> until the producer fiber has finished. Note that in our example that should
> never happen!  And even if the producer fiber did finish, it would have been
> consuming resources for nothing.

Reviewer: ```suggestion
`joinWith` or `.joinWithNever` is reached. So in our example above if
`consumerFiber` raises an error then we have no way to observe that
until the producer fiber has finished. Alarmingly, note that in our example the producer
_never_ finishes and thus the error would _never_ be observed! 
And even if the producer fiber did finish, it would have been
consuming resources for nothing.
```

Author: Good! Just please fix `then then` -> `then`.

Reviewer: Ninja'd 😉

> In contrast `parMapN` does promote the error it finds to the caller _and_ takes
> care of cancelling the remaining alive fibers. As a result `parMapN` is simpler

Reviewer: ```suggestion
In contrast `parMapN` does promote any error it finds to the caller _and_ takes
care of canceling the other running fibers. As a result `parMapN` is simpler
```
